Index: sys/dev/ic/bwfm.c
===================================================================
RCS file: /cvs/src/sys/dev/ic/bwfm.c,v
diff -u -p -r1.111 bwfm.c
--- sys/dev/ic/bwfm.c	19 Feb 2024 21:23:02 -0000	1.111
+++ sys/dev/ic/bwfm.c	5 Aug 2024 21:01:58 -0000
@@ -2867,7 +2870,7 @@ bwfm_set_key_cb(struct bwfm_softc *sc, v
 		return;
 	}
 
-	delay(100);
+	delay(500);
 
 	bwfm_fwvar_var_set_data(sc, "wsec_key", &key, sizeof(key));
 	bwfm_fwvar_var_get_int(sc, "wsec", &wsec);
