Index: Makefile
===================================================================
RCS file: /cvs/ports/devel/codechecker/Makefile,v
retrieving revision 1.8
diff -u -p -r1.8 Makefile
--- Makefile	26 Sep 2023 17:58:06 -0000	1.8
+++ Makefile	20 Oct 2023 19:57:25 -0000
@@ -2,9 +2,7 @@ COMMENT =	clang static code analyzer
 
 GH_ACCOUNT =	Ericsson
 GH_PROJECT =	codechecker
-GH_TAGNAME =	v6.21.0
-
-REVISION =	0
+GH_TAGNAME =	v6.22.2
 
 CATEGORIES =	devel
 
Index: distinfo
===================================================================
RCS file: /cvs/ports/devel/codechecker/distinfo,v
retrieving revision 1.2
diff -u -p -r1.2 distinfo
--- distinfo	22 Feb 2023 10:37:55 -0000	1.2
+++ distinfo	17 Oct 2023 20:36:51 -0000
@@ -1,10 +1,10 @@
-SHA256 (codechecker-v6.21.0/clike.min.js) = lJbma7uCu604soGpj5LJO4RxDMWUige1afF5veoZxz4=
-SHA256 (codechecker-v6.21.0/codechecker-6.21.0.tar.gz) = PDrOFSddrLgdP9fwBKlufiMP/eS/oSWnxPdpfSs2DDQ=
-SHA256 (codechecker-v6.21.0/codemirror.LICENSE) = o/L+Ksa0caqAxzfF0oPdBJvckDpzg17m1NLKwC/dU78=
-SHA256 (codechecker-v6.21.0/codemirror.min.css) = wluO/w4cnorJpS0JmcdTSYzwdb5E6u045qa4Ervfb1k=
-SHA256 (codechecker-v6.21.0/codemirror.min.js) = AqnMwc9Kk6sJShDo5QGwu2Ea9MzSR+iSWy4nJV+gRRU=
-SIZE (codechecker-v6.21.0/clike.min.js) = 18240
-SIZE (codechecker-v6.21.0/codechecker-6.21.0.tar.gz) = 11029302
-SIZE (codechecker-v6.21.0/codemirror.LICENSE) = 1107
-SIZE (codechecker-v6.21.0/codemirror.min.css) = 5731
-SIZE (codechecker-v6.21.0/codemirror.min.js) = 166305
+SHA256 (codechecker-v6.22.2/clike.min.js) = lJbma7uCu604soGpj5LJO4RxDMWUige1afF5veoZxz4=
+SHA256 (codechecker-v6.22.2/codechecker-6.22.2.tar.gz) = OMrZQFIPCS/lcWWEERHvqUWgxb+AR1ga3q1SBHv3Fs0=
+SHA256 (codechecker-v6.22.2/codemirror.LICENSE) = FopL7Mlo9QAeLuLgKRtuTaq6/BiUoRreHhHVbpYJbgc=
+SHA256 (codechecker-v6.22.2/codemirror.min.css) = wluO/w4cnorJpS0JmcdTSYzwdb5E6u045qa4Ervfb1k=
+SHA256 (codechecker-v6.22.2/codemirror.min.js) = AqnMwc9Kk6sJShDo5QGwu2Ea9MzSR+iSWy4nJV+gRRU=
+SIZE (codechecker-v6.22.2/clike.min.js) = 18240
+SIZE (codechecker-v6.22.2/codechecker-6.22.2.tar.gz) = 11056464
+SIZE (codechecker-v6.22.2/codemirror.LICENSE) = 1113
+SIZE (codechecker-v6.22.2/codemirror.min.css) = 5731
+SIZE (codechecker-v6.22.2/codemirror.min.js) = 166305
Index: pkg/PLIST
===================================================================
RCS file: /cvs/ports/devel/codechecker/pkg/PLIST,v
retrieving revision 1.1.1.1
diff -u -p -r1.1.1.1 PLIST
--- pkg/PLIST	9 Nov 2022 15:46:07 -0000	1.1.1.1
+++ pkg/PLIST	20 Oct 2023 19:58:08 -0000
@@ -64,6 +64,7 @@ share/CodeChecker/lib/python3/codechecke
 share/CodeChecker/lib/python3/codechecker_analyzer/${MODPY_PYCACHE}analyzer_context.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_analyzer/${MODPY_PYCACHE}arg.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_analyzer/${MODPY_PYCACHE}checkers.${MODPY_PYC_MAGIC_TAG}pyc
+share/CodeChecker/lib/python3/codechecker_analyzer/${MODPY_PYCACHE}compilation_database.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_analyzer/${MODPY_PYCACHE}env.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_analyzer/${MODPY_PYCACHE}gcc_toolchain.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_analyzer/${MODPY_PYCACHE}host_check.${MODPY_PYC_MAGIC_TAG}pyc
@@ -90,7 +91,6 @@ share/CodeChecker/lib/python3/codechecke
 ${MODPY_COMMENT}share/CodeChecker/lib/python3/codechecker_analyzer/analyzers/clangsa/${MODPY_PYCACHE}/
 share/CodeChecker/lib/python3/codechecker_analyzer/analyzers/clangsa/${MODPY_PYCACHE}__init__.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_analyzer/analyzers/clangsa/${MODPY_PYCACHE}analyzer.${MODPY_PYC_MAGIC_TAG}pyc
-share/CodeChecker/lib/python3/codechecker_analyzer/analyzers/clangsa/${MODPY_PYCACHE}clang_options.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_analyzer/analyzers/clangsa/${MODPY_PYCACHE}config_handler.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_analyzer/analyzers/clangsa/${MODPY_PYCACHE}ctu_autodetection.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_analyzer/analyzers/clangsa/${MODPY_PYCACHE}ctu_manager.${MODPY_PYC_MAGIC_TAG}pyc
@@ -99,7 +99,6 @@ share/CodeChecker/lib/python3/codechecke
 share/CodeChecker/lib/python3/codechecker_analyzer/analyzers/clangsa/${MODPY_PYCACHE}statistics.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_analyzer/analyzers/clangsa/${MODPY_PYCACHE}version.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_analyzer/analyzers/clangsa/analyzer.py
-share/CodeChecker/lib/python3/codechecker_analyzer/analyzers/clangsa/clang_options.py
 share/CodeChecker/lib/python3/codechecker_analyzer/analyzers/clangsa/config_handler.py
 share/CodeChecker/lib/python3/codechecker_analyzer/analyzers/clangsa/ctu_autodetection.py
 share/CodeChecker/lib/python3/codechecker_analyzer/analyzers/clangsa/ctu_manager.py
@@ -164,6 +163,7 @@ share/CodeChecker/lib/python3/codechecke
 share/CodeChecker/lib/python3/codechecker_analyzer/cmd/fixit.py
 share/CodeChecker/lib/python3/codechecker_analyzer/cmd/log.py
 share/CodeChecker/lib/python3/codechecker_analyzer/cmd/parse.py
+share/CodeChecker/lib/python3/codechecker_analyzer/compilation_database.py
 share/CodeChecker/lib/python3/codechecker_analyzer/env.py
 share/CodeChecker/lib/python3/codechecker_analyzer/gcc_toolchain.py
 share/CodeChecker/lib/python3/codechecker_analyzer/host_check.py
@@ -172,14 +172,14 @@ share/CodeChecker/lib/python3/codechecke
 share/CodeChecker/lib/python3/codechecker_analyzer/suppress_file_handler.py
 share/CodeChecker/lib/python3/codechecker_analyzer/suppress_handler.py
 share/CodeChecker/lib/python3/codechecker_api/
-share/CodeChecker/lib/python3/codechecker_api-6.51.0.dist-info/
-share/CodeChecker/lib/python3/codechecker_api-6.51.0.dist-info/INSTALLER
-share/CodeChecker/lib/python3/codechecker_api-6.51.0.dist-info/METADATA
-share/CodeChecker/lib/python3/codechecker_api-6.51.0.dist-info/RECORD
-share/CodeChecker/lib/python3/codechecker_api-6.51.0.dist-info/REQUESTED
-share/CodeChecker/lib/python3/codechecker_api-6.51.0.dist-info/WHEEL
-share/CodeChecker/lib/python3/codechecker_api-6.51.0.dist-info/direct_url.json
-share/CodeChecker/lib/python3/codechecker_api-6.51.0.dist-info/top_level.txt
+share/CodeChecker/lib/python3/codechecker_api-6.53.0.dist-info/
+share/CodeChecker/lib/python3/codechecker_api-6.53.0.dist-info/INSTALLER
+share/CodeChecker/lib/python3/codechecker_api-6.53.0.dist-info/METADATA
+share/CodeChecker/lib/python3/codechecker_api-6.53.0.dist-info/RECORD
+share/CodeChecker/lib/python3/codechecker_api-6.53.0.dist-info/REQUESTED
+share/CodeChecker/lib/python3/codechecker_api-6.53.0.dist-info/WHEEL
+share/CodeChecker/lib/python3/codechecker_api-6.53.0.dist-info/direct_url.json
+share/CodeChecker/lib/python3/codechecker_api-6.53.0.dist-info/top_level.txt
 share/CodeChecker/lib/python3/codechecker_api/Authentication_v6/
 share/CodeChecker/lib/python3/codechecker_api/Authentication_v6/__init__.py
 ${MODPY_COMMENT}share/CodeChecker/lib/python3/codechecker_api/Authentication_v6/${MODPY_PYCACHE}/
@@ -234,14 +234,14 @@ share/CodeChecker/lib/python3/codechecke
 share/CodeChecker/lib/python3/codechecker_api/codeCheckerDBAccess_v6/constants.py
 share/CodeChecker/lib/python3/codechecker_api/codeCheckerDBAccess_v6/ttypes.py
 share/CodeChecker/lib/python3/codechecker_api_shared/
-share/CodeChecker/lib/python3/codechecker_api_shared-6.51.0.dist-info/
-share/CodeChecker/lib/python3/codechecker_api_shared-6.51.0.dist-info/INSTALLER
-share/CodeChecker/lib/python3/codechecker_api_shared-6.51.0.dist-info/METADATA
-share/CodeChecker/lib/python3/codechecker_api_shared-6.51.0.dist-info/RECORD
-share/CodeChecker/lib/python3/codechecker_api_shared-6.51.0.dist-info/REQUESTED
-share/CodeChecker/lib/python3/codechecker_api_shared-6.51.0.dist-info/WHEEL
-share/CodeChecker/lib/python3/codechecker_api_shared-6.51.0.dist-info/direct_url.json
-share/CodeChecker/lib/python3/codechecker_api_shared-6.51.0.dist-info/top_level.txt
+share/CodeChecker/lib/python3/codechecker_api_shared-6.53.0.dist-info/
+share/CodeChecker/lib/python3/codechecker_api_shared-6.53.0.dist-info/INSTALLER
+share/CodeChecker/lib/python3/codechecker_api_shared-6.53.0.dist-info/METADATA
+share/CodeChecker/lib/python3/codechecker_api_shared-6.53.0.dist-info/RECORD
+share/CodeChecker/lib/python3/codechecker_api_shared-6.53.0.dist-info/REQUESTED
+share/CodeChecker/lib/python3/codechecker_api_shared-6.53.0.dist-info/WHEEL
+share/CodeChecker/lib/python3/codechecker_api_shared-6.53.0.dist-info/direct_url.json
+share/CodeChecker/lib/python3/codechecker_api_shared-6.53.0.dist-info/top_level.txt
 share/CodeChecker/lib/python3/codechecker_api_shared/__init__.py
 ${MODPY_COMMENT}share/CodeChecker/lib/python3/codechecker_api_shared/${MODPY_PYCACHE}/
 share/CodeChecker/lib/python3/codechecker_api_shared/${MODPY_PYCACHE}__init__.${MODPY_PYC_MAGIC_TAG}pyc
@@ -336,7 +336,6 @@ share/CodeChecker/lib/python3/codechecke
 share/CodeChecker/lib/python3/codechecker_common/singleton.py
 share/CodeChecker/lib/python3/codechecker_common/skiplist_handler.py
 share/CodeChecker/lib/python3/codechecker_common/tests/
-share/CodeChecker/lib/python3/codechecker_common/tests/.noserc
 share/CodeChecker/lib/python3/codechecker_common/tests/Makefile
 share/CodeChecker/lib/python3/codechecker_common/util.py
 share/CodeChecker/lib/python3/codechecker_merge_clang_extdef_mappings/
@@ -443,6 +442,12 @@ ${MODPY_COMMENT}share/CodeChecker/lib/py
 share/CodeChecker/lib/python3/codechecker_report_converter/analyzers/pylint/${MODPY_PYCACHE}__init__.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_report_converter/analyzers/pylint/${MODPY_PYCACHE}analyzer_result.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_report_converter/analyzers/pylint/analyzer_result.py
+share/CodeChecker/lib/python3/codechecker_report_converter/analyzers/roslynator/
+share/CodeChecker/lib/python3/codechecker_report_converter/analyzers/roslynator/__init__.py
+${MODPY_COMMENT}share/CodeChecker/lib/python3/codechecker_report_converter/analyzers/roslynator/${MODPY_PYCACHE}/
+share/CodeChecker/lib/python3/codechecker_report_converter/analyzers/roslynator/${MODPY_PYCACHE}__init__.${MODPY_PYC_MAGIC_TAG}pyc
+share/CodeChecker/lib/python3/codechecker_report_converter/analyzers/roslynator/${MODPY_PYCACHE}analyzer_result.${MODPY_PYC_MAGIC_TAG}pyc
+share/CodeChecker/lib/python3/codechecker_report_converter/analyzers/roslynator/analyzer_result.py
 share/CodeChecker/lib/python3/codechecker_report_converter/analyzers/sanitizers/
 share/CodeChecker/lib/python3/codechecker_report_converter/analyzers/sanitizers/__init__.py
 ${MODPY_COMMENT}share/CodeChecker/lib/python3/codechecker_report_converter/analyzers/sanitizers/${MODPY_PYCACHE}/
@@ -713,6 +718,7 @@ share/CodeChecker/lib/python3/codechecke
 share/CodeChecker/lib/python3/codechecker_server/migrations/report/versions/75ae226b5d88_review_status_for_each_report.py
 share/CodeChecker/lib/python3/codechecker_server/migrations/report/versions/82ca43f05c10_initial_schema.py
 share/CodeChecker/lib/python3/codechecker_server/migrations/report/versions/9987aa593ca7_add_codechecker_version_to_run_history.py
+share/CodeChecker/lib/python3/codechecker_server/migrations/report/versions/9d956a0fae8d_report_annotations.py
 share/CodeChecker/lib/python3/codechecker_server/migrations/report/versions/__init__.py
 ${MODPY_COMMENT}share/CodeChecker/lib/python3/codechecker_server/migrations/report/versions/${MODPY_PYCACHE}/
 share/CodeChecker/lib/python3/codechecker_server/migrations/report/versions/${MODPY_PYCACHE}080349e895d7_add_check_command_to_run_history.${MODPY_PYC_MAGIC_TAG}pyc
@@ -727,6 +733,7 @@ share/CodeChecker/lib/python3/codechecke
 share/CodeChecker/lib/python3/codechecker_server/migrations/report/versions/${MODPY_PYCACHE}75ae226b5d88_review_status_for_each_report.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_server/migrations/report/versions/${MODPY_PYCACHE}82ca43f05c10_initial_schema.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_server/migrations/report/versions/${MODPY_PYCACHE}9987aa593ca7_add_codechecker_version_to_run_history.${MODPY_PYC_MAGIC_TAG}pyc
+share/CodeChecker/lib/python3/codechecker_server/migrations/report/versions/${MODPY_PYCACHE}9d956a0fae8d_report_annotations.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_server/migrations/report/versions/${MODPY_PYCACHE}__init__.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_server/migrations/report/versions/${MODPY_PYCACHE}a24461972d2e_add_index_for_report_and_history_id_columns.${MODPY_PYC_MAGIC_TAG}pyc
 share/CodeChecker/lib/python3/codechecker_server/migrations/report/versions/${MODPY_PYCACHE}a79677f54e48_remove_db_version_table.${MODPY_PYC_MAGIC_TAG}pyc
