Index: Makefile
===================================================================
RCS file: /cvs/ports/devel/codechecker/Makefile,v
diff -u -p -r1.15 Makefile
--- Makefile	16 Jun 2025 07:47:55 -0000	1.15
+++ Makefile	18 Jun 2025 02:40:54 -0000
@@ -2,8 +2,7 @@ COMMENT =	clang static code analyzer
 
 GH_ACCOUNT =	Ericsson
 GH_PROJECT =	codechecker
-GH_TAGNAME =	v6.24.0
-REVISION =	2
+GH_TAGNAME =	v6.26.0
 
 CATEGORIES =	devel
 
Index: distinfo
===================================================================
RCS file: /cvs/ports/devel/codechecker/distinfo,v
diff -u -p -r1.3 distinfo
--- distinfo	14 Jul 2024 12:31:38 -0000	1.3
+++ distinfo	21 Jun 2025 06:17:25 -0000
@@ -1,10 +1,10 @@
-SHA256 (codechecker-v6.24.0/clike.min.js) = lJbma7uCu604soGpj5LJO4RxDMWUige1afF5veoZxz4=
-SHA256 (codechecker-v6.24.0/codechecker-6.24.0.tar.gz) = rOBtrlW7PCktUizdyQ9jBTLp1L7pOl1YhlzkHy1FF5U=
-SHA256 (codechecker-v6.24.0/codemirror.LICENSE) = FopL7Mlo9QAeLuLgKRtuTaq6/BiUoRreHhHVbpYJbgc=
-SHA256 (codechecker-v6.24.0/codemirror.min.css) = wluO/w4cnorJpS0JmcdTSYzwdb5E6u045qa4Ervfb1k=
-SHA256 (codechecker-v6.24.0/codemirror.min.js) = AqnMwc9Kk6sJShDo5QGwu2Ea9MzSR+iSWy4nJV+gRRU=
-SIZE (codechecker-v6.24.0/clike.min.js) = 18240
-SIZE (codechecker-v6.24.0/codechecker-6.24.0.tar.gz) = 11220382
-SIZE (codechecker-v6.24.0/codemirror.LICENSE) = 1113
-SIZE (codechecker-v6.24.0/codemirror.min.css) = 5731
-SIZE (codechecker-v6.24.0/codemirror.min.js) = 166305
+SHA256 (codechecker-v6.26.0/clike.min.js) = lJbma7uCu604soGpj5LJO4RxDMWUige1afF5veoZxz4=
+SHA256 (codechecker-v6.26.0/codechecker-6.26.0.tar.gz) = bjsvMpVQ6rj3x8eP2Ls6yu8yruSpf7z5xFFHwg8CqB8=
+SHA256 (codechecker-v6.26.0/codemirror.LICENSE) = FopL7Mlo9QAeLuLgKRtuTaq6/BiUoRreHhHVbpYJbgc=
+SHA256 (codechecker-v6.26.0/codemirror.min.css) = wluO/w4cnorJpS0JmcdTSYzwdb5E6u045qa4Ervfb1k=
+SHA256 (codechecker-v6.26.0/codemirror.min.js) = AqnMwc9Kk6sJShDo5QGwu2Ea9MzSR+iSWy4nJV+gRRU=
+SIZE (codechecker-v6.26.0/clike.min.js) = 18240
+SIZE (codechecker-v6.26.0/codechecker-6.26.0.tar.gz) = 11433053
+SIZE (codechecker-v6.26.0/codemirror.LICENSE) = 1113
+SIZE (codechecker-v6.26.0/codemirror.min.css) = 5731
+SIZE (codechecker-v6.26.0/codemirror.min.js) = 166305
