Index: arch/amd64/amd64/pmap.c
===================================================================
RCS file: /cvs/src/sys/arch/amd64/amd64/pmap.c,v
retrieving revision 1.164
diff -u -p -r1.164 pmap.c
--- arch/amd64/amd64/pmap.c	13 Apr 2023 15:23:21 -0000	1.164
+++ arch/amd64/amd64/pmap.c	5 May 2023 11:42:07 -0000
@@ -2497,7 +2500,7 @@ pmap_do_remove_ept(struct pmap *pmap, pa
 }
 
 int
-pmap_enter_ept(struct pmap *pmap, paddr_t gpa, paddr_t hpa, vm_prot_t prot)
+pmap_enter_ept(struct pmap *pmap, vaddr_t gpa, paddr_t hpa, vm_prot_t prot)
 {
 	uint64_t l4idx, l3idx, l2idx, l1idx;
 	pd_entry_t *pd, npte;
